_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,871 @@
+	IMPRECISERR:  0  Imprecise data bus error
+	UNSTKERR:     0  Bus fault on unstacking for a return from exception
+	STKERR:       0  Bus fault on stacking for exception entry
+	LSPERR:       0  Bus fault on floating-point lazy state preservation
+	BFARVALID:    0  Bus Fault Address Register (BFAR) valid flag
+	UNDEFINSTR:   0  Undefined instruction usage fault
+	INVSTATE:     1  Invalid state usage fault
+	INVPC:        0  Invalid PC load usage fault
+	NOCP:         0  No coprocessor usage fault.
+	UNALIGNED:    0  Unaligned access usage fault
+	DIVBYZERO:    0  Divide by zero usage fault
+```
+
+### Post-Mortem Analysis
+
+The previous two approaches are only helpful if you have a debug or physical connection to the device. Once the product has shipped and is out in the field these strategies will not help you to triage what went wrong on devices.
+
+One option is to log the fault register values at the time of crash to persistent storage and periocially collect or push the error logs. On the server side you will want to decode the register values collected and symbolicate any of the addresses collected.
+
+For those who have [Memfault](https://memfault.com/features/error-analysis.html?utm_source=interrupt&utm_medium=link&utm_campaign=cortex-m-faults) integrated, this type of data can be automatically collected, transported and deduplicated to surface the biggest problems being hit. Here's a simple example of a bad memory access from the example app we will use in the next section:
+
+![](img/cortex-m-fault/memfault-fault-analyzer.png)
+
+{: examples}
+
+## Examples
+
+For this setup we will use:
+
+- a nRF52840-DK[^7] (ARM Cortex-M4F) as our development board
+- SEGGER JLinkGDBServer[^8] as our GDB Server.
+- GCC 8.3.1 / GNU Arm Embedded Toolchain as our compiler[^9]
+- GNU make as our build system
+
+All the code can be found on the [Interrupt Github page](https://github.com/memfault/interrupt/tree/master/example/cortex-m-fault-debug) with more details in the `README` in the directory linked.
+
+In the sections below we will walk through the analysis of a couple faults.
+
+### Setup
+
+Start a GDB Server:
+
+```
+JLinkGDBServer  -if swd -device nRF52840_xxAA
+```
+
+Follow the instructions [above](#pycortex-svd-gdb-setup) to setup support for reading SVD files from GDB, build, and flask the example app:
+
+```
+$ make
+[...]
+Linking library
+Generated build/nrf52.elf
+$ arm-none-eabi-gdb-py --eval-command="target remote localhost:2331" --ex="mon reset" --ex="load" --ex="mon reset" --se=build/nrf52.elf
+$ source PyCortexMDebug/cmdebug/svd_gdb.py
+$ (gdb) svd_load cortex-m4-scb.svd
+Loading SVD file cortex-m4-scb.svd...
+(gdb)
+```
+
+The app has eight different crashes you can configure by changing `FAULT_EXAMPLE_CONFIG` at compile time or by editing the value at runtime:
+
+```
+(gdb) break main
+(gdb) continue
+(gdb) set g_crash_config=1
+(gdb) continue
+```
+
+### Exception Entry Fault
+
+Let's take a look at the first example
```

Comment:
Would be nice too reproduce the relevant snippet of C code here

Reply:
great idea, done in all the places!